ls1intum/Artemis
# Improve server side git url handling and server test improvements

We had duplicated for handling git urls in different classes. I refactored those into the new `UrlService`.
I also created the `ProgrammingExerciseTestService` to avoid duplicated functionality for programming exercise tests. At the same time, `ProgrammingExerciseResultTestService` provides testing logic for new incoming programming exercise results.
This service should be independent of the config or profile and include the logic that is needed to test programming exercises on Artemis. All logic related to external communication will be added in separate mock providers based on delegation and provided by super classes

### Checklist

### Motivation and Context
Code quality improvements and reduction of code duplication

### Steps for Testing
1. Log in to Artemis (TS1 and TS2)
2. Check that everything related to programming exercises still works

### Test Coverage
stays the same, but Jenkins and Gitlab tests will be easier to write, because we only need to specify the mocks, no additional test logic is needed.
